### PR TITLE
Ensure GitRepostory.slug is a valid Python package name

### PR DESCRIPTION
**Pull request by bryanculver**
# Closes: #DNE
# What's Changed

- Slug `GitRepository.slug` with `slugify_dashes_to_underscores`, same as the other places we use slugs for programmatic purposes
- Add test to ensure the slug is a valid Python package name
- Migration to convert existing repo slugs to valid package names

# TODO

- [x] Explanation of Change(s)
- [ ] Added change log fragment(s) (for more information see [the documentation](https://docs.nautobot.com/projects/core/en/stable/development/#creating-changelog-fragments))
- [ ] Attached Screenshots, Payload Example
- [x] Unit, Integration Tests
- [ ] Documentation Updates (when adding/changing features)
- [ ] Example Plugin Updates (when adding/changing features)
- [ ] Outline Remaining Work, Constraints from Design

### CODE REVIEW

File: nautobot/extras/models/datasources.py [hunk 1]
```diff
@@ -24,7 +24,11 @@ class GitRepository(PrimaryModel):
         max_length=100,
         unique=True,
     )
-    slug = AutoSlugField(populate_from="name")
+    slug = AutoSlugField(
+        populate_from="name",
+        help_text="Internal field name. Please use underscores rather than dashes in this key.",
+        slugify_function=slugify_dashes_to_underscores,
```
Comment (glennmatthews) [inline nautobot/extras/models/datasources.py:30]: Is this reflected in the UI or does the UI still default to dashes?

File: nautobot/extras/models/datasources.py [hunk 2]
```diff
@@ -24,7 +24,11 @@ class GitRepository(PrimaryModel):
...
+    )
 
     remote_url = models.URLField(
         max_length=255,
```
Comment (glennmatthews) [inline nautobot/extras/models/datasources.py:34]: Comparing against #3426, isn't this missing `clean()` logic to enforce validation of the slug?

Comment (bryanculver) [reply nautobot/extras/models/datasources.py:34]: My bad, I assumed the field did the validation. Adding.

**Review by glennmatthews — COMMENTED** [review]
Actually, we may also want to pull in the code from custom-fields that makes `slug` immutable once set without deleting/recreating the repository? Since changing it will affect import paths for jobs?

**Comment by glennmatthews** [comment]
> > Actually, we may also want to pull in the code from custom-fields that makes `slug` immutable once set without deleting/recreating the repository? Since changing it will affect import paths for jobs?
> 
> What if I refactor my jobs? Do I have to delete and re-add the repo?

I wouldn't think so? I'm working on handling this in my current work.

> Also it would be pretty fragile to do full package name import since the author of the job may not have access to the GitRepository configuration.

Jobs shouldn't do absolute imports within a Git repository, but on the Nautobot side we have to have an absolute path to import the jobs from. That path *would* change when re-slugging a GitRepository.

**Review by glennmatthews — APPROVED** [review]

File: nautobot/project-static/js/forms.js [hunk 3]
```diff
@@ -4,14 +4,16 @@
 
 // Slugify
 function slugify(s, num_chars) {
-    s = s.replace(/[^\-\.\w\s]/g, '');          // Remove unneeded chars
+    s = s.replace(/[^_A-Za-z0-9\-\s\.]/g, '');  // Remove non-ascii chars
     s = s.replace(/^[\s\.]+|[\s\.]+$/g, '');    // Trim leading/trailing spaces
     s = s.replace(/[\-\.\s]+/g, '-');           // Convert spaces and decimals to hyphens
```
[the remark this answers is not in the document]
Comment (bryanculver) [reply nautobot/project-static/js/forms.js:9]: Trying to closely match how the Python slugging is happening which does it in two steps: Replace multiple instances of these characters (in Django `slugify`), then replace the `-` with `_` in our `slugify_dashes_to_underscores`.

**Review by timizuoebideri1 — APPROVED** [review]

**Pull request merged**